src/tracerunner.rs

```diff
@@ -0,0 +1,153 @@
+
+use std::rc::Rc;
+
+use kaktus::PushPop;
+
+use super::{TraceInstruction, Comp, Value, Interpreter, CallFrame, Func};
+use recovery::Guard;
+use traits::vec::ConvertingStack;
+
+
+pub struct Runner<'a, 'b: 'a> {
+    pub trace: &'a [TraceInstruction],
+    pub stack: Vec<Value>,
+    pub locals: Vec<Value>,
+    pub interp: &'a mut Interpreter<'b>,
+}
+
+impl<'a, 'b> Runner<'a, 'b> {
+    pub fn new(interp: &'a mut Interpreter<'b>,
+               trace: &'a [TraceInstruction],
+               n_locals: usize)
+               -> Self {
+        // we have to copy over current stack frame from interpreter
+        let mut locals = vec![Value::Null; n_locals];
+        {
+            let interp_locals = &interp.frames.last().unwrap().locals;
+            for idx in 0..interp_locals.len() {
+                locals[idx] = interp_locals[idx].clone();
+            }
+        }
+
+        Runner {
+            interp: interp,
+            trace: trace,
+            stack: Vec::new(),
+            locals: locals,
+        }
+    }
+
+    pub fn run(&mut self) -> (Rc<Func>, usize) {
+        use TraceInstruction::*;
+
+        let mut pc = 0;
+        loop {
+            let instr = &self.trace[pc];
+            pc = (pc + 1) % self.trace.len();
+
+            info!("TEXEC: {:?}", instr);
+
+            match *instr {
+                Add => self.add(),
+                Cmp(how) => self.cmp(how),
+
+                Load(idx) => self.load(idx),
+                Store(idx) => self.store(idx),
+                Const(val) => self.stack.push_from(val),
+
+                ArrayGet => self.array_get(),
+
+                Clone => (),
+
+                Guard(ref guard) => {
+                    match self.guard(guard) {
+                        Ok(_) => (),
+                        Err(recovery) => return recovery,
+                    }
+                }
+
+                _ => unimplemented!(),
+                // Array(usize),
+                // Push,
+                // Print,
+                // Len,
+            }
+
+        }
+    }
+
+    // XXX: return None guard succeeds
+    fn guard(&mut self, guard: &Guard) -> Result<(), (Rc<Func>, usize)> {
+        let got = self.stack.pop_into::<bool>();
+        if got == guard.condition {
+            Ok(())
+        } else {
+            self.recover(guard);
+            Err((guard.frame.func.clone(), guard.pc))
+        }
+    }
+
+    /// the following things have to be recovered
+    /// * stack-frames (call-frames)
+    /// * value stack (essentially bool which caused guard to fail)
+    fn recover(&mut self, guard: &Guard) {
```

Comment:
I don't really know much about black-holing, so you are kinda on your own here ;)